Topic-4-Javascript-Desing-Patterns/4/4.js

```diff
@@ -0,0 +1,84 @@
+class EventEmiter {
+    constructor () {
+        this.events = {};
+    }
+    on (eventName , callback) {
+        if ( this.events[eventName]) {
+            this.events[eventName].push(callback);
+        } else {
+            this.events[eventName] = [callback];
+        }        
+    }
+    emit (eventName) {
+        this.events[eventName].forEach(callback =>
+            callback(`The ${eventName} has been emited`));
+    }
+    off (eventName , callback) {
+        let newEvents = this.events[eventName].filter(c => c !== callback);
+        this.events[eventName] = newEvents;
+    }
+    
+}
+
+class Movie extends EventEmiter {
+
+    constructor (title,year,duration) {
+        super();
+        this.title = title;
+        this.year = year;
+        this.duration = duration;
+        this.cast = [];
+    }
+    play() {
+        this.emit("play");
+    }
+    pause() {
+        this.emit("pause");
+    }
+    resume() {
+        this.emit("resume");
+    }
+    addCast(cast) {
+        this.cast = cast;
+        console.log("cast added " + this.cast.name);
+        
+        for (let i=0; i<cast.length; i++) {
```

Comment:
This logic is weird:

1. What if `cast` is an `Actor object`? 
2. If I send an Array of Actors, you run add `n` times this: `this.cast = cast`, maybe using `push` to add the actors?
3. This logic doesn't fully add actors because it erases the current actors.